---
title: "Tohidul: baseline 3: Priming gene expression"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
theme: cosmo
---

Source codes: https://github.com/markziemann/tohidul_rnaseq

## Background

Here we have n=3 control (H2O; "H") and n=3 seaweed based fertiliser treatments:

* S80 ANDP

* S94 DP

* S93 AN

Reads underwent quality trimming using Skewer (Jiang et al, 2014).
I mapped the reads to the Arabidopsis transcriptome (TAIR10/Ensembl47) using Kallisto (Bray et al, 2016).
Expression counts were loaded into R and then DE analysis was performed with DESeq2 (Love et al, 2014).
Enrichment analysis was performed using Plant Reactome genesets with the Mitch package (Kaspi & Ziemann 2020).

```{r,pkg}

suppressPackageStartupMessages({
  library("reshape2")
  library("gplots")
  library("DESeq2")
  library("mitch")
  library("kableExtra")
  library("eulerr")
  library("UpSetR")
  library("vioplot")

})

```

## Load data

Here we load the data in from the aligner.
Take note that the columns are arranged in order.

```{r,import1}

tmp <- read.table("3col.tsv.gz")
x <- as.data.frame(acast(tmp, V2~V1, value.var="V3"))
x$gene <- sapply(strsplit(rownames(x),"\\."),"[[",1)
xx <- aggregate(. ~ gene, x, sum)
rownames(xx) <- xx$gene
xx$gene = NULL
xx <- xx[,order(colnames(xx))]

```

## Sample sheet

The sample sheet is read in and put in order as well.

```{r,samplesheet1}

ss <- read.table("samplesheet.tsv",header=TRUE)
ss <- ss[order(ss$Run),]
ss$label <- gsub("-0hpi","",ss$SampleName)
# check that names are in order
colnames(xx) == ss$Run
# change data header
colnames(xx) <- ss$label
kbl(ss) %>% kable_styling()

```

## MDS

MDS is just like PCA.
The more similar (correlated) the data sets are the closer they will appear on the scatterplot.

```{r,mds1,fig.height=8,fig.width=8}

cols <- as.numeric(factor(ss$Treatment))
plot(cmdscale(dist(t(xx))), xlab="Coordinate 1", ylab="Coordinate 2", 
  col=cols, cex=4 , main="MDS plot")
text(cmdscale(dist(t(xx))), labels=ss$label )

#colfunc <- colorRampPalette(c("white", "yellow", "orange" ,  "red", "darkred"))
colfunc <- colorRampPalette(c("white","blue"))

heatmap.2(cor(xx,method="pearson"),col=colfunc(25),trace="none",
  scale="none",margin=c(10,10),main="Pearson correlation")

heatmap.2(cor(xx,method="spearman"),col=colfunc(25),trace="none",
  scale="none",margin=c(10,10),main="Spearman correlation")
 
```


## Functions

```{r,defunc,fig.height=8,fig.width=8}

run_de <- function(ss,xx){
xx <- xx[which(rowMeans(xx)>10),]
y <- round(xx)
# MDS
mds <- cmdscale(dist(t(y)))
XMAX=max(mds[,1])*1.1
XMIN=min(mds[,1])*1.1
plot( mds , xlab="Coordinate 1", ylab="Coordinate 2",
  type = "n" , xlim=c(XMIN,XMAX),main="MDS plot",bty="n")
text(mds, labels=colnames(y) )
# DE
dds <- DESeqDataSetFromMatrix(countData=y, colData = ss, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de <- de[order(de$pvalue),]
up <- rownames(subset(de, log2FoldChange>0 & padj<0.05 ))
dn <- rownames(subset(de, log2FoldChange<0 & padj<0.05 ))
str(up)
str(dn)
# MA plot
sig <-subset(de, padj < 0.05 )
GENESUP <- length(up)
GENESDN <- length(dn)
SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down")
ns <-subset(de, padj > 0.05 )
plot(log2(de$baseMean),de$log2FoldChange,
     xlab="log2 basemean", ylab="log2 foldchange",
     pch=19, cex=0.5, col="dark gray",
     main="smear plot")
points(log2(sig$baseMean),sig$log2FoldChange,
       pch=19, cex=0.5, col="red")
mtext(SUBHEADER)
# heatmap
yn <- y/colSums(y)*1000000
yf <- yn[which(rownames(yn) %in% rownames(de)[1:50]),]
mycols <- gsub("0","yellow",ss$trt)
mycols <- gsub("1","orange",mycols)
colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(  as.matrix(yf), col=colfunc(25),scale="row",
    ColSideColors =mycols ,trace="none",
    margin = c(10,10), cexRow=0.6, cexCol=0.8 , main="Top 50 genes by p-val")
mtext("yellow=ctrl, orange=trt")
return(de)
}


mitch_barplot <- function(res){
  sig <- res$enrichment_result
  sig <- head( sig[order(sig$pANOVA),] ,30)
  sig <- sig[order(sig$s.dist),]
  par(mar=c(3,25,1,1)); barplot(sig$s.dist,horiz=TRUE,las=2,cex.names = 0.6,cex.axis = 0.6,
    names.arg=sig$set,main="Enrichment score") ;grid()
}

```


## Split data into contrast groups

```{r,split1}

ss80 <- subset(ss,Treatment=="Seaweed_16280"|Treatment=="Water")
ss80$trt <- as.numeric(grepl("Seaweed",ss80$Treatment))
xx80 <- xx[,which(colnames(xx) %in% ss80$label)]

```

## DE

Here, were using DESeq2 to perform differential expression analysis to understand gene expression changes caused by the different formulations.
Enrichment analysis is performed at the end using mitch with all 4 timepoints.

```{r,de1,fig.height=8,fig.width=8}

de80 <- run_de(ss80,xx80)

de80_top <- head(as.data.frame(de80),40)
kbl(de80_top[order(-de80_top$log2FoldChange),]) %>% kable_styling()

write.table(de80,file="de80.tsv",quote=FALSE,sep="\t")

de80_up <- rownames(subset(de80,padj<0.05&log2FoldChange>0))
de80_dn <- rownames(subset(de80,padj<0.05&log2FoldChange<0))


```

## Pathway analysis

Mapman pathways last modified in 2012 and used in the previous RNA-seq analysis.

```{r,mapman,fig.height=8,fig.width=8}

genesets <- gmt_import("../baseline/ref/Ath_AGI_LOCUS_TAIR10_Aug2012.txt.gmt")

gt <- read.table("../baseline/ref/Arabidopsis_thaliana.TAIR10.46.geneaccession2symbol.tsv",
    fill=TRUE) 

m <- mitch_import(x=as.data.frame(de80),DEtype="deseq2",geneTable=gt)

# significance
res_sig <- mitch_calc(x=m,genesets=genesets,priority="significance")

top <- head(subset(res_sig$enrichment_result,p.adjustANOVA<0.05),50)
kbl(head(top,30)) %>% kable_styling()

mitch_barplot(res_sig)



# effect size
res_eff <- mitch_calc(x=m,genesets=genesets,priority="effect")

top <- head(subset(res_eff$enrichment_result,p.adjustANOVA<0.05),50)
kbl(head(top,30)) %>% kable_styling()

mitch_barplot(res_eff)

```


Gene sets of interest


```{r,genesetsofinterest, fig.height=8,fig.width=8}

mysets <- c("TRANSPORT_CALCIUM",
  "REDOX_THIOREDOXIN_PDIL",
  "CELL_WALL_CELLULOSE_SYNTHESIS",
  "PROTEIN_TARGETING_CHLOROPLAST",
  "MITOCHONDRIAL_ELECTRON_TRANSPORT_/_ATP_SYNTHESIS_CYTOCHROME_C")

mysets <- genesets[which(names(genesets) %in% mysets)]
 
myres <- mitch_calc(x=m,genesets=mysets,priority="effect")

par(mar=c(5,20,5,2))
vioplot(myres$detailed_sets,horizontal=TRUE,las=2,side="right",cex.axis=0.7)


```


## Mitch reports

```{r,mitch_reports}

unlink("mapman_report_de80_sig.html")
capture.output(
    mitch_report(res_sig,outfile=paste("mapman_report_de80_sig.html"))
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

unlink("mapman_report_de80_eff.html")
capture.output(
    mitch_report(res_eff,outfile=paste("mapman_report_de80_eff.html"))
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

```

## Make PDFs

```{r,pdf}

pdf("de80.pdf")
de80 <- run_de(ss80,xx80)
mitch_barplot(res_eff)
par(mar=c(5,20,5,2))
vioplot(myres$detailed_sets,horizontal=TRUE,las=2,side="right",cex.axis=0.7)
dev.off()

```

## Session information

So you know what version of R and packages was used.

```{r,session}

sessionInfo()

```

## References

Bray NL, Pimentel H, Melsted P, Pachter L. Near-optimal probabilistic RNA-seq quantification [published correction appears in Nat Biotechnol. 2016 Aug 9;34(8):888]. Nat Biotechnol. 2016;34(5):525-527. doi:10.1038/nbt.3519

Jiang H, Lei R, Ding SW, Zhu S. Skewer: a fast and accurate adapter trimmer for next-generation sequencing paired-end reads. BMC Bioinformatics. 2014;15:182. Published 2014 Jun 12. doi:10.1186/1471-2105-15-182

Love MI, Huber W, Anders S. Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biol. 2014;15(12):550. doi:10.1186/s13059-014-0550-8

Kaspi A, Ziemann M. mitch: multi-contrast pathway enrichment for multi-omics and single-cell profiling data. BMC Genomics. 2020;21(1):447. Published 2020 Jun 29. doi:10.1186/s12864-020-06856-9

