---
title: "Tohidul: does Seasol prime gene expression"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
theme: cosmo
---

Source codes: https://github.com/markziemann/tohidul_rnaseq

## Background


Here we have n=6 control (control "H") and n=6 seaweed (SW).
We have separate leaf and root samples.

Reads underwent quality trimming using Skewer (Jiang et al, 2014).
I mapped the reads to the Arabidopsis transcriptome (TAIR10/Ensembl47) using Kallisto (Bray et al, 2016).
Expression counts were loaded into R and then DE analysis was performed with DESeq2 (Love et al, 2014).
Enrichment analysis was performed using Plant Reactome genesets with the Mitch package (Kaspi & Ziemann 2020).

```{r,pkg}

suppressPackageStartupMessages({
  library("reshape2")
  library("gplots")
  library("DESeq2")
  library("mitch")
  library("kableExtra")
  library("eulerr")
  library("UpSetR")
  library("vioplot")
  library(ggplot2)
})

```

## Load data

Here we load the data in from the aligner.
Take note that the columns are arranged in order.

```{r,import1}

tmp <- read.table("3col.tsv.gz")
x <- as.data.frame(acast(tmp, V2~V1, value.var="V3"))
x$gene <- sapply(strsplit(rownames(x),"\\."),"[[",1)
xx <- aggregate(. ~ gene, x, sum)
rownames(xx) <- xx$gene
xx$gene = NULL
xx <- xx[,order(colnames(xx))]
write.table(xx,file="seasol_2021-03_genecounts.tsv",sep="\t",quote=FALSE)

```

## Sample sheet

The sample sheet is read in and put in order as well.

```{r,samplesheet1}

ss <- read.table("samplesheet.tsv",header=TRUE,sep="\t")
ss <- ss[order(ss$sample),]

```

## MDS

MDS is just like PCA.
The more similar (correlated) the data sets are the closer they will appear on the scatterplot.

```{r,mds1,fig.height=8,fig.width=8}

cols <- as.numeric(factor(ss$tissue))
cols <- gsub(1,"pink",cols)
cols <- gsub(2,"lightblue",cols)

plot(cmdscale(dist(t(xx))), xlab="Coordinate 1", ylab="Coordinate 2", 
  col=cols, cex=4 , pch=19, main="MDS plot all samples",bty="n")

plot(cmdscale(dist(t(xx))), xlab="Coordinate 1", ylab="Coordinate 2",
  col=cols, cex=4 , pch=19, main="MDS plot all samples",bty="n")
text(cmdscale(dist(t(xx))), labels=colnames(xx) )

leaf <- xx[,grep("-L",colnames(xx))] 

ssl <- subset(ss,tissue=="leaf")

cols <- as.numeric(factor(ssl$treatment))
cols <- gsub(1,"pink",cols)
cols <- gsub(2,"lightblue",cols)

plot(cmdscale(dist(t(leaf))), xlab="Coordinate 1", ylab="Coordinate 2",
  col=cols, cex=4 , main="MDS plot leaf samples",pch=19,bty="n")
text(cmdscale(dist(t(leaf))), labels=colnames(leaf) )

root <- xx[,grep("-R",colnames(xx))]

ssr <- subset(ss,tissue=="root")

cols <- as.numeric(factor(ssr$treatment))
cols <- gsub(1,"pink",cols)
cols <- gsub(2,"lightblue",cols)

plot(cmdscale(dist(t(root))), xlab="Coordinate 1", ylab="Coordinate 2",
  col=cols, cex=4 , main="MDS plot root samples",pch=19,bty="n")
text(cmdscale(dist(t(root))), labels=colnames(root) )

colfunc <- colorRampPalette(c("white","blue"))

heatmap.2(cor(xx,method="pearson"),col=colfunc(25),trace="none",
  scale="none",margin=c(10,10),main="Pearson correlation all samples")

heatmap.2(cor(xx,method="spearman"),col=colfunc(25),trace="none",
  scale="none",margin=c(10,10),main="Spearman correlation all samples")
 
heatmap.2(cor(leaf,method="pearson"),col=colfunc(25),trace="none",
  scale="none",margin=c(10,10),main="Pearson correlation leaf samples")

heatmap.2(cor(leaf,method="spearman"),col=colfunc(25),trace="none",
  scale="none",margin=c(10,10),main="Spearman correlation leaf samples")

heatmap.2(cor(root,method="pearson"),col=colfunc(25),trace="none",
  scale="none",margin=c(10,10),main="Pearson correlation root samples")

heatmap.2(cor(root,method="spearman"),col=colfunc(25),trace="none",
  scale="none",margin=c(10,10),main="Spearman correlation root samples")

```


## Functions

```{r,defunc,fig.height=8,fig.width=8}

run_de <- function(ss,xx){
xx <- xx[which(rowMeans(xx)>10),]
y <- round(xx)
# MDS
mds <- cmdscale(dist(t(y)))
XMAX=max(mds[,1])*1.1
XMIN=min(mds[,1])*1.1
plot( mds , xlab="Coordinate 1", ylab="Coordinate 2",
  type = "n" , xlim=c(XMIN,XMAX),main="MDS plot",bty="n")
text(mds, labels=colnames(y) )
# DE
dds <- DESeqDataSetFromMatrix(countData=y, colData = ss, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de <- de[order(de$pvalue),]
up <- rownames(subset(de, log2FoldChange>0 & padj<0.05 ))
dn <- rownames(subset(de, log2FoldChange<0 & padj<0.05 ))
str(up)
str(dn)
# MA plot
sig <-subset(de, padj < 0.05 )
GENESUP <- length(up)
GENESDN <- length(dn)
SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down")
ns <-subset(de, padj > 0.05 )
plot(log2(de$baseMean),de$log2FoldChange,
     xlab="log2 basemean", ylab="log2 foldchange",
     pch=19, cex=0.5, col="dark gray",
     main="smear plot")
points(log2(sig$baseMean),sig$log2FoldChange,
       pch=19, cex=0.5, col="red")
mtext(SUBHEADER)
# heatmap
yn <- y/colSums(y)*1000000
yf <- yn[which(rownames(yn) %in% rownames(de)[1:50]),]
mycols <- gsub("0","yellow",ss$trt)
mycols <- gsub("1","orange",mycols)
colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(  as.matrix(yf), col=colfunc(25),scale="row",
    ColSideColors =mycols ,trace="none",
    margin = c(10,10), cexRow=0.6, cexCol=0.8 , main="Top 50 genes by p-val")
mtext("yellow=ctrl, orange=trt")
return(de)
}


mitch_barplot <- function(res){
  sig <- res$enrichment_result
  sig <- head( sig[order(sig$pANOVA),] ,30)
  sig <- sig[order(sig$s.dist),]
  par(mar=c(3,25,1,1)); barplot(sig$s.dist,horiz=TRUE,las=2,cex.names = 0.6,cex.axis = 0.6,
    names.arg=sig$set,main="Enrichment score") ;grid()
}

mitch_bubbleplot <- function(res,n) {
  top <- head(res$enrichment_result,n)
  top <- top[order(top$s.dist),]
  top$set <- factor(top$set, levels = top$set[order(top$s.dist)])
  ggplot(top, aes(s.dist, set, size = setSize)) + geom_point(aes(colour=-log10(top$p.adjustANOVA)))
}

```


## Split data into contrast groups

```{r,split1}

l3 <- leaf[,grep("3D",colnames(leaf))]
ssl3 <- subset(ssl,days==3)
ssl3$trt <- as.numeric(ssl3$treatment == "seaweed")

l5 <- leaf[,grep("5D",colnames(leaf))] 
ssl5 <- subset(ssl,days==5)
ssl5$trt <- as.numeric(ssl5$treatment == "seaweed")

r3 <- root[,grep("3D",colnames(root))]
ssr3 <- subset(ssr,days==3)
ssr3$trt <- as.numeric(ssr3$treatment == "seaweed")

r5 <- root[,grep("5D",colnames(root))]
ssr5 <- subset(ssr,days==5)
ssr5$trt <- as.numeric(ssr5$treatment == "seaweed")

```

## DE

Here, were using DESeq2 to perform differential expression analysis to understand gene expression changes
caused by seaweed treatment.
Enrichment analysis is performed with mitch.

```{r,de1,fig.height=8,fig.width=8}

del3 <- run_de(ssl3,l3)
del5 <- run_de(ssl5,l5)
der3 <- run_de(ssr3,r3)
der5 <- run_de(ssr5,r5)

del3_top <- head(as.data.frame(del3),40)
kbl(del3_top[order(del3_top$log2FoldChange),]) %>% kable_paper("hover", full_width = F)

del5_top <- head(as.data.frame(del5),40)
kbl(del5_top[order(del5_top$log2FoldChange),]) %>% kable_paper("hover", full_width = F)

der3_top <- head(as.data.frame(der3),40)
kbl(der3_top[order(der3_top$log2FoldChange),]) %>% kable_paper("hover", full_width = F)

der5_top <- head(as.data.frame(der5),40)
kbl(der5_top[order(der5_top$log2FoldChange),]) %>% kable_paper("hover", full_width = F)

write.table(del3,file="del3.tsv",quote=FALSE,sep="\t")
write.table(del5,file="del5.tsv",quote=FALSE,sep="\t")
write.table(der3,file="der3.tsv",quote=FALSE,sep="\t")
write.table(der5,file="der5.tsv",quote=FALSE,sep="\t")

del3_up <- rownames(subset(del3,padj<0.05&log2FoldChange>0))
del3_dn <- rownames(subset(del3,padj<0.05&log2FoldChange<0))

del5_up <- rownames(subset(del5,padj<0.05&log2FoldChange>0))
del5_dn <- rownames(subset(del5,padj<0.05&log2FoldChange<0))

der3_up <- rownames(subset(der3,padj<0.05&log2FoldChange>0))
der3_dn <- rownames(subset(der3,padj<0.05&log2FoldChange<0))

der5_up <- rownames(subset(der5,padj<0.05&log2FoldChange>0))
der5_dn <- rownames(subset(der5,padj<0.05&log2FoldChange<0))

up <- sapply(list(del3_up,del5_up,der3_up,der5_up),length)
dn <- sapply(list(del3_dn,del5_dn,der3_dn,der5_dn),length)

res_summary <- data.frame(up,dn,row.names=c("L3","L5","R3","R5"))

res_summary$de <- c("Leaf 3d","Leaf 5d","Root 3d","Root 5d")

kbl(res_summary) %>% kable_paper("hover", full_width = F)

barplot(res_summary$up,ylim=c(-2500,2500), axes=TRUE, main="number of DEGs up/down")

barplot(-res_summary$dn, add = TRUE, names.arg = rownames(res_summary), axes = FALSE)

```

### Venn diagram plots of overlap

```{r,venn1,fig.height=8,fig.width=8}

# upregulated genes
v1 <- list("leaf 3d up"=del3_up, "leaf 5d up"=del5_up, 
  "root 3d up" =der3_up , "root 5d up"=der5_up)

plot(euler(v1),quantities = TRUE)

intersect(del3_up,del5_up)
intersect(der3_up,der3_up)

# downregulated genes
v2 <- list("leaf 3d dn"=del3_dn, "leaf 5d dn"=del5_dn, 
  "root 3d dn" =der3_dn , "root 5d dn"=der5_dn)

plot(euler(v2),quantities = TRUE)

intersect(del3_dn,del5_dn)
intersect(der3_dn,der3_dn)

```

## DEG summary

Here is a heatmap of all the differentially expressed genes.

```{r,bigheat,fig.height=8,fig.width=8}

degs <- unique(c(del3_up, del3_dn,
  del5_up, del5_dn,
  der3_up, der3_dn,
  der5_up, der5_dn))

xxx <- xx/colSums(xx)*1e6
deg_mx <- as.matrix(xxx[which(rownames(xxx) %in% degs),])
colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(deg_mx,col=colfunc(25),cexCol=0.8,trace="none",scale="row",margin=c(10,10),main="all samples")


degs_leaf <- unique(c(del3_up, del3_dn, del5_up, del5_dn))
leafx <-  leaf/colSums(leaf)*1e6
deg_mx <- as.matrix(leaf[which(rownames(leaf) %in% degs_leaf),])
colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(deg_mx,col=colfunc(25),cexCol=0.8,trace="none",scale="row",margin=c(10,10),main="leaf samples")

degs_root <- unique(c(der3_up, der3_dn, der5_up, der5_dn))
rootx <-  root/colSums(root)*1e6
deg_mx <- as.matrix(root[which(rownames(root) %in% degs_root),])
colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(deg_mx,col=colfunc(25),cexCol=0.8,trace="none",scale="row",margin=c(10,10),main="root samples")

```

## Pathway analysis

Mapman pathways last modified in 2012 and used in the previous RNA-seq analysis.
First I'll strt with unidimensional analysis.

```{r,unidimensional,fig.height=8,fig.width=8}

genesets <- gmt_import("../../ref/Ath_AGI_LOCUS_TAIR10_Aug2012.txt.gmt")

gt <- read.table("../../ref/Arabidopsis_thaliana.TAIR10.46.geneaccession2symbol.tsv",
    fill=TRUE) 
gt$V1 <- sapply(strsplit(gt$V1,"\\."),"[[",1)
gt <- unique(gt)


m <- mitch_import(x=as.data.frame(del3), DEtype="deseq2")
del3m <- mitch_calc(x=m,genesets=genesets,priority="effect",minsetsize=5)
mitch_barplot(del3m)
mitch_bubbleplot(del3m,30)

unlink("del3m_mitch.html")
capture.output(
    mitch_report(del3m, outfile="del3m_mitch.html"),
    file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

m <- mitch_import(x=as.data.frame(del5),DEtype="deseq2")
del5m <- mitch_calc(x=m,genesets=genesets,priority="effect",minsetsize=5)
mitch_barplot(del5m)
mitch_bubbleplot(del5m,30)

unlink("del5m_mitch.html")
capture.output(
    mitch_report(del5m, outfile="del5m_mitch.html"),
    file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

m <- mitch_import(x=as.data.frame(der3),DEtype="deseq2")
der3m <- mitch_calc(x=m,genesets=genesets,priority="effect",minsetsize=5)
mitch_barplot(der3m)
mitch_bubbleplot(der3m,30)

unlink("der3m_mitch.html")
capture.output(
    mitch_report(der3m, outfile="der3m_mitch.html"),
    file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

m <- mitch_import(x=as.data.frame(der5),DEtype="deseq2")
der5m <- mitch_calc(x=m,genesets=genesets,priority="effect",minsetsize=5)
mitch_barplot(der5m)
mitch_bubbleplot(der5m,30)

unlink("der5m_mitch.html")
capture.output(
    mitch_report(der5m, outfile="der5m_mitch.html"),
    file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

```

Next, multidimensional analysis.

```{r,multidimensional,fig.height=8,fig.width=8}

l <- list("del3"=del3,"del5"=del5,
  "der3"=der3, "der5"=der5)

m <- mitch_import(x=l,DEtype="deseq2")

# effect size
res <- mitch_calc(x=m,genesets=genesets,priority="effect",minsetsize=5)

top <- subset(res$enrichment_result,p.adjustMANOVA<0.05)
kbl(head(top,30)) %>% kable_paper("hover", full_width = F)
rownames(top) <- top[,1]
top <- top[1:50,4:7]
colfunc <- colorRampPalette(c("blue", "white", "red"))
colnames(top) <- gsub("s\\.","",colnames(top))

heatmap.2(  as.matrix(top), col=colfunc(25),
    scale="none",Colv=FALSE,trace="none", dendrogram="row",
    margins = c(5,25), cexCol=0.8 , cexRow=0.7, main="Top genes sets by effect size (FDR<0.05)")

unlink("mapman_report_eff.html")
capture.output(
    mitch_report(res,outfile=paste("mapman_report_eff.html"))
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

```

## Focus on ROS

DEL3 contrast

```{r,customsets_del3,fig.height=8,fig.width=8}

customsets <- c("redox.ascorbate.and.glutathione.ascorbate",
  "misc.peroxidases",
  "redox.dismutases.and.catalases",
  "redox.glutaredoxins")

customsets <- genesets[which(names(genesets) %in% customsets)]

m <- mitch_import(x=as.data.frame(del3), DEtype="deseq2")
del3m <- mitch_calc(x=m,genesets=customsets,priority="effect",minsetsize=5)

del3m$enrichment_result %>%
  kbl() %>%
  kable_paper("hover", full_width = F)

mitch_barplot(del3m)
vioplot(del3m$detailed_sets,horizontal=TRUE,las=1) ; grid()
mitch_bubbleplot(del3m,30)

#unlink("del3m_redox.html")
#capture.output(
#    mitch_report(del3m, outfile="del3m_redox.html"),
#    file = "/dev/null", append = FALSE,
#    type = c("output", "message"), split = FALSE)

```

DEL5

```{r,customsets_del5,fig.height=8,fig.width=8}

m <- mitch_import(x=as.data.frame(del5), DEtype="deseq2",geneTable)
del5m <- mitch_calc(x=m,genesets=customsets,priority="effect",minsetsize=5)

del5m$enrichment_result %>%
  kbl() %>%
  kable_paper("hover", full_width = F)

mitch_barplot(del5m)
vioplot(del5m$detailed_sets,horizontal=TRUE,las=1) ; grid()
mitch_bubbleplot(del5m,30)

#unlink("del5m_redox.html")
#capture.output(
#    mitch_report(del5m, outfile="del5m_redox.html"),
#    file = "/dev/null", append = FALSE,
#    type = c("output", "message"), split = FALSE)

```

DER3

```{r,customsets_der3,fig.height=8,fig.width=8}

m <- mitch_import(x=as.data.frame(der3), DEtype="deseq2")
der3m <- mitch_calc(x=m,genesets=customsets,priority="effect",minsetsize=5)

der3m$enrichment_result %>%
  kbl() %>%
  kable_paper("hover", full_width = F)

mitch_barplot(der3m)
vioplot(der3m$detailed_sets,horizontal=TRUE,las=1) ; grid()
mitch_bubbleplot(der3m,30)

unlink("der3m_redox.html")
capture.output(
    mitch_report(der3m, outfile="der3m_redox.html"),
    file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

```

DER5

```{r,customsets_der5,fig.height=8,fig.width=8}

m <- mitch_import(x=as.data.frame(der5), DEtype="deseq2")
der5m <- mitch_calc(x=m,genesets=customsets,priority="effect",minsetsize=5)

der5m$enrichment_result %>%
  kbl() %>%
  kable_paper("hover", full_width = F)

mitch_barplot(der5m)
vioplot(der5m$detailed_sets,horizontal=TRUE,las=1) ; grid()
mitch_bubbleplot(der5m,30)

unlink("der5m_redox.html")
capture.output(
    mitch_report(der5m, outfile="der5m_redox.html"),
    file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

```

## Session information

So you know what version of R and packages was used.

```{r,session}

sessionInfo()

```

## References

Bray NL, Pimentel H, Melsted P, Pachter L. Near-optimal probabilistic RNA-seq quantification [published correction appears in Nat Biotechnol. 2016 Aug 9;34(8):888]. Nat Biotechnol. 2016;34(5):525-527. doi:10.1038/nbt.3519

Jiang H, Lei R, Ding SW, Zhu S. Skewer: a fast and accurate adapter trimmer for next-generation sequencing paired-end reads. BMC Bioinformatics. 2014;15:182. Published 2014 Jun 12. doi:10.1186/1471-2105-15-182

Love MI, Huber W, Anders S. Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biol. 2014;15(12):550. doi:10.1186/s13059-014-0550-8

Kaspi A, Ziemann M. mitch: multi-contrast pathway enrichment for multi-omics and single-cell profiling data. BMC Genomics. 2020;21(1):447. Published 2020 Jun 29. doi:10.1186/s12864-020-06856-9

