---
title: "Tohidul effect of Seasol on baseline gene expression - timecourse"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
theme: cosmo
---

## Introduction

In this analysis I am performing differential expression and pathway analysis on seasol and control plants at all timepoints (0, 3, 6, 12 and 24 hrs).

## Methods

The RNA-seq data is already deposited at SRA, so the raw data has been processed and incorporated into the DEE2 database, obtainable with the SRA project accession number SRP253869 (Ziemann et al, 2019).
After obtaining processed data from DEE2 using the R/Bioconductor package, it will undergo differential expression analysis at each timepoint using DESeq2 (Love et al, 2014). 
Next, the differential expression profile underwent enrichment analysis of Mapman gene sets using the mitch software package (Kaspi & Ziemann 2020).
Mitch allows to perform multi-dimensional enrichment analysis, so control-treatment enrichment analysis ca be performed at all timepoints at once.

```{r , libraries }
suppressPackageStartupMessages({
    library("getDEE2")
    library("reshape2")
    library("gplots")
    library("DESeq2")
    library("mitch")
    library("edgeR")
    library("RColorBrewer")
    library("gplots")
  })
```

## Set up the sample sheet

Using sample information from dee2.io.

```{r pressure }
md <- getDEE2Metadata("athaliana")

md <- md[which(md$SRP_accession=="SRP253869"),]

ctrl <- grep("H-",md$Library_name)

s80 <- grep("S80",md$Library_name)

md <- md[c(ctrl,s80),]

timepoint <- sapply(strsplit(md$Library_name, "-"),"[[",2)

timepoint <- sapply(strsplit(timepoint, "_"),"[[",1)

md$timepoint <- timepoint

md$trt <- as.numeric(grepl("S80",md$Library_name))

md
```



## Fetch dataset

Obtaining the RNA expression data from dee2.io.

```{r, fetch}

# fetch the expresion data
x <- getDEE2(species="athaliana", SRRvec = md$SRR_accession , legacy = TRUE)

# collapse tx wise expression counts to genes
x <- Tx2Gene(x)

```

Here, I'm running an MDS plot to see the overall variation in the expression data.

```{r,mds}
samplegroups <- factor(sapply(strsplit(md$Library_name,"_"),"[[",1))

colour_palette <- brewer.pal(n = length(levels(samplegroups)), name = "Paired")

colours <- colour_palette[as.integer(factor(samplegroups))]

plot(1,axes = FALSE,xlab="",ylab="",main="MDS by ART type")

legend("center",legend=levels(samplegroups),pch=16,cex=1.2,col=colour_palette)

mydist <- plotMDS(x$Tx2Gene, labels=colnames(x$Tx2Gene),col=colours,main="MDS plot")

```

``` {r, split}

# split into different objects for analysis
ss0 <- md[which(md$timepoint=="0hpi"),]
x0 <- x$Tx2Gene[,which(colnames(x$Tx2Gene) %in% ss0$SRR_accession)]
x0 <- x0[which(rowMeans(x0)>10),]
x0 <- round(x0)

ss3 <- md[which(md$timepoint=="3hpi"),]
x3 <- x$Tx2Gene[,which(colnames(x$Tx2Gene) %in% ss3$SRR_accession)]
x3 <- x3[which(rowMeans(x3)>10),]
x3 <- round(x3)

ss6 <- md[which(md$timepoint=="6hpi"),]
x6 <- x$Tx2Gene[,which(colnames(x$Tx2Gene) %in% ss6$SRR_accession)]
x6 <- x6[which(rowMeans(x6)>10),]
x6 <- round(x6)

ss12 <- md[which(md$timepoint=="12hpi"),]
x12 <- x$Tx2Gene[,which(colnames(x$Tx2Gene) %in% ss12$SRR_accession)]
x12 <- x12[which(rowMeans(x12)>10),]
x12 <- round(x12)

ss24 <- md[which(md$timepoint=="24hpi"),]
x24 <- x$Tx2Gene[,which(colnames(x$Tx2Gene) %in% ss24$SRR_accession)]
x24 <- x24[which(rowMeans(x24)>10),]
x24 <- round(x24)

```

## Differential expression

Now it is time to run differential expression analysis.

```{r, de}
# t=0
dds <- DESeqDataSetFromMatrix(countData=x0, colData = ss0, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de0 <- as.data.frame(de[order(de$pvalue),])
head(de0,20)


# t=3
dds <- DESeqDataSetFromMatrix(countData=x3, colData = ss3, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de3 <- as.data.frame(de[order(de$pvalue),])
head(de3,20)

# t=6
dds <- DESeqDataSetFromMatrix(countData=x6, colData = ss6, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de6 <- as.data.frame(de[order(de$pvalue),])
head(de6,20)

# t=12
dds <- DESeqDataSetFromMatrix(countData=x12, colData = ss12, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de12 <- as.data.frame(de[order(de$pvalue),])
head(de12,20)

# t=24
dds <- DESeqDataSetFromMatrix(countData=x24, colData = ss24, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de24 <- as.data.frame(de[order(de$pvalue),])
head(de24,20)

```

Now run an MDS plot to visualise the gene expression differences at each timepoint.

```{r, maplot}

maplot <- function(de,contrast_name) {
  sig <-subset(de, padj < 0.05 )
  up <-rownames(subset(de, padj < 0.05 & log2FoldChange > 0))
  dn <-rownames(subset(de, padj < 0.05 & log2FoldChange < 0))
  GENESUP <- length(up)
  GENESDN <- length(dn)
  SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down")
  ns <-subset(de, padj > 0.05 )
  plot(log2(de$baseMean),de$log2FoldChange, 
       xlab="log2 basemean", ylab="log2 foldchange",
       pch=19, cex=0.5, col="dark gray",
       main=contrast_name, cex.main=0.7)
  points(log2(sig$baseMean),sig$log2FoldChange,
         pch=19, cex=0.5, col="red")
  mtext(SUBHEADER,cex = 0.7)
}

maplot(de0, contrast_name = "0hpi")
maplot(de3, contrast_name = "3hpi")
maplot(de6, contrast_name = "6hpi")
maplot(de12, contrast_name = "12hpi")
maplot(de24, contrast_name = "24hpi")

```

## Enrichment analysis

```{r, mitch}
gsets <- gmt_import("Ath_AGI_LOCUS_TAIR10_Aug2012.txt.gmt")

xl <- list("de0"=de0,"de3"=de3,"de6"=de6,"de12"=de12,"de24"=de24)

y <- mitch_import(xl,DEtype = "DESeq2")

# prioritisation by effect size
capture.output(
    res <- mitch_calc(y,gsets,priority="effect")
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

head(res$enrichment_result,20)

capture.output(
    mitch_plots(res,"timecourse_mitch_eff_plots.pdf")
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

unlink("timecourse_mitch_eff_report.html")
capture.output(
    mitch_report(res,"timecourse_mitch_eff_report.html")
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

```

## References

Ziemann M, Kaspi A, El-Osta A. Digital expression explorer 2: a repository of uniformly processed RNA sequencing data. Gigascience. 2019;8(4):giz022. doi:10.1093/gigascience/giz022

Love MI, Huber W, Anders S. Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biol. 2014;15(12):550. doi:10.1186/s13059-014-0550-8

Kaspi A, Ziemann M. mitch: multi-contrast pathway enrichment for multi-omics and single-cell profiling data. BMC Genomics. 2020;21(1):447. Published 2020 Jun 29. doi:10.1186/s12864-020-06856-9

## Session information

For reproducibility.

```{r, sessioninfo}
sessionInfo()
```

