---
title: "Tohidul effect of Seasol on baseline gene expression - timecourse"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
theme: cosmo
---

## Introduction

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents.

```{r , libraries, echo=FALSE}
suppressPackageStartupMessages({
    library("getDEE2")
    library("reshape2")
    library("gplots")
    library("DESeq2")
    library("mitch")
})
```

## Set up the sample sheet

Using sample information from dee2.io.

```{r pressure, echo=FALSE}
md <- getDEE2Metadata("athaliana")

md <- md[which(md$SRP_accession=="SRP253869"),]

ctrl <- grep("H-",md$Library_name)

s80 <- grep("S80",md$Library_name)

md <- md[c(ctrl,s80),]

timepoint <- sapply(strsplit(md$Library_name, "-"),"[[",2)

timepoint <- sapply(strsplit(timepoint, "_"),"[[",1)

md$timepoint <- timepoint

md$trt <- as.numeric(grepl("S80",md$Library_name))

md
```



## Fetch dataset

Obtaining the RNA expression data from dee2.io.

```{r, fetch}
# fetch the expresion data
x <- getDEE2(species="athaliana", SRRvec = md$SRR_accession , legacy = TRUE)

# collapse tx wise expression counts to genes
x <- Tx2Gene(x)

# split into different objects for analysis
ss0 <- md[which(md$timepoint=="0hpi"),]
x0 <- x$Tx2Gene[,which(colnames(x$Tx2Gene) %in% ss0$SRR_accession)]
x0 <- round(x0)

ss3 <- md[which(md$timepoint=="3hpi"),]
x3 <- x$Tx2Gene[,which(colnames(x$Tx2Gene) %in% ss3$SRR_accession)]
x3 <- round(x3)

ss6 <- md[which(md$timepoint=="6hpi"),]
x6 <- x$Tx2Gene[,which(colnames(x$Tx2Gene) %in% ss6$SRR_accession)]
x6 <- round(x6)

ss12 <- md[which(md$timepoint=="12hpi"),]
x12 <- x$Tx2Gene[,which(colnames(x$Tx2Gene) %in% ss12$SRR_accession)]
x12 <- round(x12)

ss24 <- md[which(md$timepoint=="24hpi"),]
x24 <- x$Tx2Gene[,which(colnames(x$Tx2Gene) %in% ss24$SRR_accession)]
x24 <- round(x24)


```

## Differential expression

```{r, de}
# t=0
dds <- DESeqDataSetFromMatrix(countData=x0, colData = ss0, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de0 <- as.data.frame(de[order(de$pvalue),])
head(de0,20)

# t=3
dds <- DESeqDataSetFromMatrix(countData=x3, colData = ss3, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de3 <- as.data.frame(de[order(de$pvalue),])
head(de3,20)

# t=6
dds <- DESeqDataSetFromMatrix(countData=x6, colData = ss6, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de6 <- as.data.frame(de[order(de$pvalue),])
head(de6,20)

# t=12
dds <- DESeqDataSetFromMatrix(countData=x12, colData = ss12, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de12 <- as.data.frame(de[order(de$pvalue),])
head(de12,20)

# t=24
dds <- DESeqDataSetFromMatrix(countData=x24, colData = ss24, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de24 <- as.data.frame(de[order(de$pvalue),])
head(de24,20)

```
## Enrichment analysis
```{r, mitch}
gsets <- gmt_import("Ath_AGI_LOCUS_TAIR10_Aug2012.txt.gmt")

xl <- list("de0"=de0,"de3"=de3,"de6"=de6,"de12"=de12,"de24"=de24)

y <- mitch_import(xl,DEtype = "DESeq2")

# prioritisation by significance
capture.output(
    res <- mitch_calc(y,gsets,priority="significance")
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

head(res$enrichment_result,20)

capture.output(
    mitch_plots(res,"timecourse_mitch_sig_plots.pdf")
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

unlink("timecourse_mitch_sig_report.html")
capture.output(
    mitch_report(res,"timecourse_mitch_sig_report.html")
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

# prioritisation by effect size
capture.output(
    res <- mitch_calc(y,gsets,priority="effect")
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

head(res$enrichment_result,20)

capture.output(
    mitch_plots(res,"timecourse_mitch_eff_plots.pdf")
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

unlink("timecourse_mitch_eff_report.html")
capture.output(
    mitch_report(res,"timecourse_mitch_eff_report.html")
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)


```


## Session information

For reproducibility.

```{r, sessioninfo}
sessionInfo()
```

