---
title: "Tohidul: timecourse effect of Seasol on gene expression"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
theme: cosmo
---

Source codes: https://github.com/markziemann/tohidul_rnaseq

## Background

Here we have n=3 control (H2O; "H") and n=3 seaweed based fertiliser treatment ("80").
Arabidopsis RNA samples.
Reads underwent quality trimming using Skewer (Jiang et al, 2014).
I mapped the reads to the Arabidopsis transcriptome (TAIR10/Ensembl47) using Kallisto (Bray et al, 2016).
Expression counts were loaded into R and then DE analysis was performed with DESeq2 (Love et al, 2014).
Enrichment analysis was performed using Plant Reactome genesets with the Mitch package (Kaspi & Ziemann 2020).

```{r,pkg}
suppressPackageStartupMessages({
    library("reshape2")
    library("gplots")
    library("DESeq2")
    library("mitch")
})
```

## Load data

Here we load the data in from the aligner.

```{r,import}
tmp <- read.table("3col.tsv")
x <- as.data.frame(acast(tmp, V2~V1, value.var="V3"))
x$gene <- sapply(strsplit(rownames(x),"\\."),"[[",1)
xx <- aggregate(. ~ gene, x, sum)
rownames(xx) <- xx$gene
xx$gene = NULL
```

## Sample sheet

```{r,samplesheet}
ss <- data.frame(colnames(xx))
rownames(ss) <- ss[,1]
ss$trt <- as.numeric(grepl("ANDP",ss[,1]))
ss[,1]=NULL

```

## MDS

MDS is just like PCA.
The more similar (correlated) the data sets are the closer they will appear on the scatterplot.

```{r,mds,fig.height=8,fig.width=8}
plot(cmdscale(dist(t(xx))), xlab="Coordinate 1", ylab="Coordinate 2", 
  type = "n" , main="MDS plot")
text(cmdscale(dist(t(xx))), labels=colnames(xx) )
 
```

## Split data into timepoints

```{r,split}
ss0 <- ss[grep("-0",rownames(ss)),,drop=FALSE]
ss1 <- ss[grep("-1",rownames(ss)),,drop=FALSE]
ss3 <- ss[grep("-3",rownames(ss)),,drop=FALSE]
ss5 <- ss[grep("-5",rownames(ss)),,drop=FALSE]
xx0 <- xx[,grep("-0",colnames(xx))]
xx1 <- xx[,grep("-1",colnames(xx))]
xx3 <- xx[,grep("-3",colnames(xx))]
xx5 <- xx[,grep("-5",colnames(xx))]
xx0 <- xx0[which(rowMeans(xx0)>10),]
xx1 <- xx1[which(rowMeans(xx1)>10),]
xx3 <- xx3[which(rowMeans(xx3)>10),]
xx5 <- xx5[which(rowMeans(xx5)>10),]

```

## DE

Here, were using DESeq2 to perform differential expression analysis at the different time points.
Enrichment analysis is performed at the end using mitch with all 4 timepoints.

### Time=0

```{r,de0,fig.height=8,fig.width=8}
ss <- ss0
y <- xx0

y <- round(y)
dds <- DESeqDataSetFromMatrix(countData=y, colData = ss, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de <- de[order(de$pvalue),]
head(de)
write.table(de,file="t0.tsv",sep="\t",quote=FALSE)
de0<-de

# define up and down-regulated gene lists
up <- rownames(subset(de, log2FoldChange>0 & padj<0.05 ))
dn <- rownames(subset(de, log2FoldChange<0 & padj<0.05 ))
str(up)
str(dn)

# MA plot
sig <-subset(de, padj < 0.05 )
GENESUP <- length(up)
GENESDN <- length(dn)
SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down")
ns <-subset(de, padj > 0.05 )
plot(log2(de$baseMean),de$log2FoldChange, 
     xlab="log2 basemean", ylab="log2 foldchange",
     pch=19, cex=0.5, col="dark gray",
     main="smear plot")
points(log2(sig$baseMean),sig$log2FoldChange,
       pch=19, cex=0.5, col="red")
mtext(SUBHEADER)

# heatmap
yn <- y/colSums(y)*1000000
yf <- yn[which(rownames(yn) %in% rownames(de)[1:50]),]

mycols <- gsub("0","yellow",ss$trt)
mycols <- gsub("1","orange",mycols)

colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(  as.matrix(yf), col=colfunc(25),scale="row", ColSideColors =mycols ,trace="none",
    margins = c(10,10), cexRow=0.6, main="Top 50 genes by p-val")
mtext("yellow=ctrl, orange=trt")

```

### Time=1

```{r,de1,fig.height=8,fig.width=8}
ss <- ss1
y <- xx1

y <- round(y)
dds <- DESeqDataSetFromMatrix(countData=y, colData = ss, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de <- de[order(de$pvalue),]
head(de)
write.table(de,file="t1.tsv",sep="\t",quote=FALSE)
de1<-de

# define up and down-regulated gene lists
up <- rownames(subset(de, log2FoldChange>0 & padj<0.05 ))
dn <- rownames(subset(de, log2FoldChange<0 & padj<0.05 ))
str(up)
str(dn)

# MA plot
sig <-subset(de, padj < 0.05 )
GENESUP <- length(up)
GENESDN <- length(dn)
SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down")
ns <-subset(de, padj > 0.05 )
plot(log2(de$baseMean),de$log2FoldChange,
     xlab="log2 basemean", ylab="log2 foldchange",
     pch=19, cex=0.5, col="dark gray",
     main="smear plot")
points(log2(sig$baseMean),sig$log2FoldChange,
       pch=19, cex=0.5, col="red")
mtext(SUBHEADER)

# heatmap
yn <- y/colSums(y)*1000000
yf <- yn[which(rownames(yn) %in% rownames(de)[1:50]),]

mycols <- gsub("0","yellow",ss$trt)
mycols <- gsub("1","orange",mycols)

colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(  as.matrix(yf), col=colfunc(25),scale="row", ColSideColors =mycols ,trace="none",
    margins = c(10,10), cexRow=0.6, main="Top 50 genes by p-val")
mtext("yellow=ctrl, orange=trt")

```

### Time=3

```{r,de3,fig.height=8,fig.width=8}
ss <- ss3
y <- xx3

y <- round(y)
dds <- DESeqDataSetFromMatrix(countData=y, colData = ss, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de <- de[order(de$pvalue),]
head(de)
write.table(de,file="t3.tsv",sep="\t",quote=FALSE)
de3<-de

# define up and down-regulated gene lists
up <- rownames(subset(de, log2FoldChange>0 & padj<0.05 ))
dn <- rownames(subset(de, log2FoldChange<0 & padj<0.05 ))
str(up)
str(dn)

# MA plot
sig <-subset(de, padj < 0.05 )
GENESUP <- length(up)
GENESDN <- length(dn)
SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down")
ns <-subset(de, padj > 0.05 )
plot(log2(de$baseMean),de$log2FoldChange,
     xlab="log2 basemean", ylab="log2 foldchange",
     pch=19, cex=0.5, col="dark gray",
     main="smear plot")
points(log2(sig$baseMean),sig$log2FoldChange,
       pch=19, cex=0.5, col="red")
mtext(SUBHEADER)

# heatmap
yn <- y/colSums(y)*1000000
yf <- yn[which(rownames(yn) %in% rownames(de)[1:50]),]

mycols <- gsub("0","yellow",ss$trt)
mycols <- gsub("1","orange",mycols)

colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(  as.matrix(yf), col=colfunc(25),scale="row", ColSideColors =mycols ,trace="none",
    margins = c(10,10), cexRow=0.6, main="Top 50 genes by p-val")
mtext("yellow=ctrl, orange=trt")

```

### Time=5

```{r,de5,fig.height=8,fig.width=8}
ss <- ss5
y <- xx5

y <- round(y)
dds <- DESeqDataSetFromMatrix(countData=y, colData = ss, design = ~ trt)
dds <- DESeq(dds)
de <- DESeq2::results(dds)
de <- de[order(de$pvalue),]
head(de)
write.table(de,file="t5.tsv",sep="\t",quote=FALSE)
de5<-de

# define up and down-regulated gene lists
up <- rownames(subset(de, log2FoldChange>0 & padj<0.05 ))
dn <- rownames(subset(de, log2FoldChange<0 & padj<0.05 ))
str(up)
str(dn)

# MA plot
sig <-subset(de, padj < 0.05 )
GENESUP <- length(up)
GENESDN <- length(dn)
SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down")
ns <-subset(de, padj > 0.05 )
plot(log2(de$baseMean),de$log2FoldChange,
     xlab="log2 basemean", ylab="log2 foldchange",
     pch=19, cex=0.5, col="dark gray",
     main="smear plot")
points(log2(sig$baseMean),sig$log2FoldChange,
       pch=19, cex=0.5, col="red")
mtext(SUBHEADER)

# heatmap
yn <- y/colSums(y)*1000000
yf <- yn[which(rownames(yn) %in% rownames(de)[1:50]),]

mycols <- gsub("0","yellow",ss$trt)
mycols <- gsub("1","orange",mycols)

colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(  as.matrix(yf), col=colfunc(25),scale="row", ColSideColors =mycols ,trace="none",
    margins = c(10,10), cexRow=0.6, main="Top 50 genes by p-val")
mtext("yellow=ctrl, orange=trt")

```

## Pathway analysis

Mapman pathways last modified in 2012 and used in the previous RNA-seq analysis.

```{r,mapman,fig.height=8,fig.width=8}
genesets <- gmt_import("../ref/Ath_AGI_LOCUS_TAIR10_Aug2012.txt.gmt")
gt <- read.table("../ref/Arabidopsis_thaliana.TAIR10.46.geneaccession2symbol.tsv",
    fill=TRUE) 

l <- list("T0"=de0,"T1"=de1,"T3"=de3,"T5"=de5)

m <- mitch_import(x=l,DEtype="deseq2",geneTable=gt)

capture.output(
    res <- mitch_calc(x=m,genesets=genesets,priority="effect")
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

top <- subset(res$enrichment_result,p.adjustMANOVA<0.05)
head(top,30)
rownames(top) <- top[,1]
top <- top[,4:7]
colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(  as.matrix(top), col=colfunc(25),scale="row", trace="none",
    margins = c(10,25), cexRow=0.5, main="Top genes sets")

unlink("mapman_report.html")
capture.output(
    mitch_report(res,outfile=paste("mapman_report.html"))
    , file = "/dev/null", append = FALSE,
    type = c("output", "message"), split = FALSE)

```

## Session information

So you know what version of R and packages was used.

```{r,session}
sessionInfo()
```

## References

Bray NL, Pimentel H, Melsted P, Pachter L. Near-optimal probabilistic RNA-seq quantification [published correction appears in Nat Biotechnol. 2016 Aug 9;34(8):888]. Nat Biotechnol. 2016;34(5):525-527. doi:10.1038/nbt.3519

Jiang H, Lei R, Ding SW, Zhu S. Skewer: a fast and accurate adapter trimmer for next-generation sequencing paired-end reads. BMC Bioinformatics. 2014;15:182. Published 2014 Jun 12. doi:10.1186/1471-2105-15-182

Love MI, Huber W, Anders S. Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biol. 2014;15(12):550. doi:10.1186/s13059-014-0550-8

Kaspi A, Ziemann M. mitch: multi-contrast pathway enrichment for multi-omics and single-cell profiling data. BMC Genomics. 2020;21(1):447. Published 2020 Jun 29. doi:10.1186/s12864-020-06856-9

